engine/src/main/java/org/hibernate/validator/internal/engine/validationcontext/AbstractValidationContext.java

```diff
@@ -343,31 +343,22 @@ private boolean isAlreadyValidatedForPath(Object value, PathImpl path) {
 			return false;
 		}
 
+		if ( path.isRootPath() ) {
+			return true;
+		}
+
+		// Since this isAlreadyValidatedForPath(..) is only applicable for an object that is about to be cascaded into,
+		// it means that the new path we are testing cannot be a root path; also since we are cascading into inner
+		// objects, i.e. going further from the object tree root, it means that the new path cannot be shorter than
+		// the ones we've already encountered.
 		for ( PathImpl p : pathSet ) {
-			if ( path.isRootPath() || p.isRootPath() || isSubPathOf( path, p ) || isSubPathOf( p, path ) ) {
+			if ( p.isSubPathOrContains( path ) ) {
```

Comment:
This was more or less by try and error. If I don't check both sides, it will loose Ops/s (substantial)  I think the issue stems from the Set, in which there might be a longer path appearing earlier in the iteration.

If I understand the code correctly, the set contains all possible paths to a property. In the debugger, it is visible, that a node-equals or path equals will not only compare the property string, it will also consider other properties. An effect is, that there are path that have an equal toString representation but are not equal. If this makes sense is out of my understanding of the codebase but I find it suspicious.

Reply:
Reagrding the isSubPathOf(), you are right, as it is not required for the moment, I can remove it.

Reply:
I cleaned up the commit history and reduced it to two commits. One is the fix in the original location, the other commit moves the code to PathImpl.

Reply:
Thanks!

> I think the issue stems from the Set, in which there might be a longer path appearing earlier in the iteration.

Hmm, right ... I suppose it could happen if there's a cyclic object graph like we have in `CascadedValidation$CascadedValidationState`. (that a longer path is already in the set)

I was thinking that maybe we could try to optimize it a bit more, but going with the way you've done it now seems safer as it matches the previous logic while still improving the throughput a bit 😃